bug: test_exp_inversions_lower failure

I was running the test suite locally. Got one failure in `test_exp`. Here's the test output:

```sh
Ⓟ starkware ~/projects/aura/contracts dev/gage_gate_v2 ✗ ➭ pytest -n auto tests
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.9.6, pytest-7.1.1, pluggy-1.0.0
rootdir: /Users/m/projects/aura/contracts/tests, configfile: pytest.ini
plugins: web3-5.28.0, xdist-2.5.0, forked-1.4.0, asyncio-0.15.1, typeguard-2.12.1, profiling-1.7.0, hypothesis-6.45.1
gw0 [143] / gw1 [143] / gw2 [143] / gw3 [143] / gw4 [143] / gw5 [143] / gw6 [143] / gw7 [143] / gw8 [143] / gw9 [143]
...........................................................F...................................................................................                                                                [100%]
====================================================================================================== FAILURES ======================================================================================================
_____________________________________________________________________________________________ test_exp_inversions_lower ______________________________________________________________________________________________
[gw7] darwin -- Python 3.9.6 /Users/m/.pyenv/versions/3.9.6/envs/starkware/bin/python3.9

deploy_test_contract = <starkware.starknet.testing.contract.StarknetContract object at 0x12c7d97c0>

    @settings(max_examples=50, deadline=None)
>   @given(val=st_lower_range)

tests/lib/test_exp.py:73:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Users/m/.pyenv/versions/3.9.6/envs/starkware/lib/python3.9/site-packages/pytest_asyncio/plugin.py:195: in inner
    _loop.run_until_complete(task)
/Users/m/.pyenv/versions/3.9.6/lib/python3.9/asyncio/base_events.py:642: in run_until_complete
    return future.result()
tests/lib/test_exp.py:84: in test_exp_inversions_lower
    assert_equalish(result * inverse_result, Decimal(1), Decimal("0.06"))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = Decimal('0.833333333333333333689552216752837045'), b = Decimal('1'), error = Decimal('0.06')

    def assert_equalish(a: Decimal, b: Decimal, error=ERROR_MARGIN):
>       assert abs(a - b) <= error
E       AssertionError

tests/utils.py:149: AssertionError
----------------------------------------------------------------------------------------------------- Hypothesis -----------------------------------------------------------------------------------------------------
Falsifying example: test_exp_inversions_lower(
    val=-39654772204664767312,
    deploy_test_contract=<starkware.starknet.testing.contract.StarknetContract at 0x12c7d97c0>,
)
=============================================================================================== Hypothesis Statistics ================================================================================================
lib/test_exp.py::test_exp_fail:

  - during reuse phase (0.06 seconds):
    - Typical runtimes: ~ 62ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (0.48 seconds):
    - Typical runtimes: 41-66 ms, ~ 1% in data generation
    - 9 passing examples, 0 failing examples, 0 invalid examples

  - Stopped because settings.max_examples=10


lib/test_exp.py::test_exp_pass_lower:

  - during reuse phase (0.05 seconds):
    - Typical runtimes: ~ 48ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (3.59 seconds):
    - Typical runtimes: 44-81 ms, ~ 2% in data generation
    - 49 passing examples, 0 failing examples, 1 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_floor:

  - during reuse phase (0.03 seconds):
    - Typical runtimes: ~ 27ms, ~ 1% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (1.63 seconds):
    - Typical runtimes: 20-43 ms, ~ 4% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_ceil:

  - during reuse phase (0.04 seconds):
    - Typical runtimes: ~ 40ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (1.70 seconds):
    - Typical runtimes: 23-43 ms, ~ 3% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_add_sub[test_add-add]:

  - during generate phase (2.58 seconds):
    - Typical runtimes: 0-63 ms, ~ 35% in data generation
    - 50 passing examples, 0 failing examples, 46 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_add_sub[test_sub-sub]:

  - during generate phase (2.29 seconds):
    - Typical runtimes: 0-47 ms, ~ 27% in data generation
    - 50 passing examples, 0 failing examples, 30 invalid examples

  - Stopped because settings.max_examples=50


lib/test_exp.py::test_exp_inversions_lower:

  - during reuse phase (0.77 seconds):
    - Typical runtimes: 112-231 ms, ~ 0% in data generation
    - 0 passing examples, 5 failing examples, 0 invalid examples
    - Found 1 distinct error in this phase

  - during shrink phase (8.25 seconds):
    - Typical runtimes: 100-421 ms, ~ 0% in data generation
    - 51 passing examples, 1 failing examples, 0 invalid examples
    - Tried 52 shrinks of which 0 were successful

  - Stopped because nothing left to do


shared/test_wad_ray.py::test_add_sub_unsigned[test_add_unsigned-add]:

  - during generate phase (2.04 seconds):
    - Typical runtimes: 24-54 ms, ~ 1% in data generation
    - 50 passing examples, 0 failing examples, 0 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_add_sub_unsigned[test_sub_unsigned-sub]:

  - during generate phase (1.74 seconds):
    - Typical runtimes: 24-45 ms, ~ 1% in data generation
    - 50 passing examples, 0 failing examples, 0 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_mul_div_signed[test_wsigned_div-floordiv-1000000000000000000-wad]:

  - during generate phase (2.43 seconds):
    - Typical runtimes: 0-53 ms, ~ 36% in data generation
    - 50 passing examples, 0 failing examples, 60 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_mul_div_signed[test_rmul-mul-1000000000000000000000000000-ray]:

  - during generate phase (2.48 seconds):
    - Typical runtimes: 0-55 ms, ~ 37% in data generation
    - 50 passing examples, 0 failing examples, 64 invalid examples

  - Stopped because settings.max_examples=50


lib/test_exp.py::test_exp_sum_lower:

  - during reuse phase (0.16 seconds):
    - Typical runtimes: ~ 156ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (10.14 seconds):
    - Typical runtimes: 0-408 ms, ~ 23% in data generation
    - 49 passing examples, 0 failing examples, 25 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_mul_div_signed[test_rsigned_div-floordiv-1000000000000000000000000000-ray]:

  - during generate phase (2.31 seconds):
    - Typical runtimes: 0-53 ms, ~ 32% in data generation
    - 50 passing examples, 0 failing examples, 50 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_div_unsigned[test_wunsigned_div-floordiv-1000000000000000000-wad]:

  - during generate phase (2.24 seconds):
    - Typical runtimes: 0-47 ms, ~ 32% in data generation
    - 50 passing examples, 0 failing examples, 46 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_div_unsigned[test_runsigned_div-floordiv-1000000000000000000000000000-ray]:

  - during generate phase (1.91 seconds):
    - Typical runtimes: 0-48 ms, ~ 34% in data generation
    - 50 passing examples, 0 failing examples, 46 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_div_unsigned[test_runsigned_div_unchecked-floordiv-1000000000000000000000000000-ray]:

  - during generate phase (2.22 seconds):
    - Typical runtimes: 0-46 ms, ~ 30% in data generation
    - 50 passing examples, 0 failing examples, 40 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_mul_div_signed[test_wmul-mul-1000000000000000000-wad]:

  - during generate phase (2.23 seconds):
    - Typical runtimes: 0-49 ms, ~ 37% in data generation
    - 50 passing examples, 0 failing examples, 56 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_wadray_conversions_pass[test_to_wad-int-to_wad-wad]:

  - during reuse phase (0.03 seconds):
    - Typical runtimes: ~ 26ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (1.93 seconds):
    - Typical runtimes: 21-45 ms, ~ 4% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_div_unsigned[test_wunsigned_div_unchecked-floordiv-1000000000000000000-wad]:

  - during generate phase (2.01 seconds):
    - Typical runtimes: 0-40 ms, ~ 35% in data generation
    - 50 passing examples, 0 failing examples, 56 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_wadray_conversions_pass[test_wad_to_felt-to_wad-int-wad]:

  - during reuse phase (0.03 seconds):
    - Typical runtimes: ~ 25ms, ~ 1% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (2.08 seconds):
    - Typical runtimes: 23-68 ms, ~ 1% in data generation
    - 49 passing examples, 0 failing examples, 0 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_wadray_conversions_pass[test_wad_to_ray-int-wad_to_ray-ray]:

  - during reuse phase (0.03 seconds):
    - Typical runtimes: ~ 24ms, ~ 1% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (2.07 seconds):
    - Typical runtimes: 22-66 ms, ~ 3% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_wadray_conversions_pass[test_wad_to_ray_unchecked-int-wad_to_ray-ray]:

  - during reuse phase (0.04 seconds):
    - Typical runtimes: ~ 38ms, ~ 6% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (1.78 seconds):
    - Typical runtimes: 22-47 ms, ~ 4% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_uint_conversion_pass[test_to_uint-int-to_uint-uint]:

  - during generate phase (2.14 seconds):
    - Typical runtimes: 23-54 ms, ~ 3% in data generation
    - 50 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


shared/test_wad_ray.py::test_uint_conversion_pass[test_from_uint-to_uint-int-wad]:

  - during generate phase (2.23 seconds):
    - Typical runtimes: 23-61 ms, ~ 3% in data generation
    - 50 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


lib/test_exp.py::test_exp_pass_upper:

  - during reuse phase (0.06 seconds):
    - Typical runtimes: ~ 55ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (3.37 seconds):
    - Typical runtimes: 41-76 ms, ~ 3% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


lib/test_exp.py::test_exp_inversions_upper:

  - during reuse phase (0.11 seconds):
    - Typical runtimes: ~ 106ms, ~ 0% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (6.82 seconds):
    - Typical runtimes: 91-362 ms, ~ 3% in data generation
    - 49 passing examples, 0 failing examples, 2 invalid examples

  - Stopped because settings.max_examples=50


lib/test_exp.py::test_exp_sum_upper:

  - during reuse phase (0.19 seconds):
    - Typical runtimes: ~ 190ms, ~ 5% in data generation
    - 1 passing examples, 0 failing examples, 0 invalid examples

  - during generate phase (9.98 seconds):
    - Typical runtimes: 0-447 ms, ~ 21% in data generation
    - 49 passing examples, 0 failing examples, 19 invalid examples

  - Stopped because settings.max_examples=50


============================================================================================== short test summary info ===============================================================================================
FAILED tests/lib/test_exp.py::test_exp_inversions_lower - AssertionError
=============================================================================== 1 failed, 142 passed, 11 warnings in 498.20s (0:08:18) ===============================================================================
```

--- a/tests/lib/test_exp.py
+++ b/tests/lib/test_exp.py
@@ -74,6 +74,7 @@ async def test_exp_pass_lower(deploy_test_contract, val):
 @given(val=st_lower_range)
 @example(val=to_wad(25))
 @example(val=-to_wad(40))
+@example(val=-39654772204664767312)
 @pytest.mark.asyncio
 async def test_exp_inversions_lower(deploy_test_contract, val):
     contract = deploy_test_contract
@@ -82,7 +83,12 @@ async def test_exp_inversions_lower(deploy_test_contract, val):
     inverse_result = Decimal(from_wad((await contract.get_exp(signed_int_to_felt(-val)).execute()).result.res))
 
     # Precision starts getting pretty bad with all these multiplications and divisions
-    assert_equalish(result * inverse_result, Decimal(1), Decimal("0.06"))
+    if val < -to_wad(39):
+        error_margin = Decimal("0.17")
+    else:
+        error_margin = Decimal("0.06")
+
+    assert_equalish(result * inverse_result, Decimal(1), error_margin)
 
 
 # This tests that exp(x+y) = exp(x)*exp(y)

